lib/utils.js

```diff
@@ -75,19 +78,60 @@ exports.console = Object.freeze(jpmConsole);
 /**
  * Returns the `package.json` manifest as an object
  * from the `cwd`, or `null` if not found.
+ * If you pass in an optional XPI filename, the manifest will be returned
+ * from this directory after a temporary XPI extraction.
  *
- * @return {Object}
+ * @param {String} optional XPI filename
+ * @return {Promise} resolves to a manifest object
  */
-function getManifest () {
+function getManifest (xpiPath) {
   return when.promise(function(resolve, reject) {
-    var cwd = process.cwd();
-    var json = path.join(cwd, "package.json");
-    var manifest = {};
-    try {
-      manifest = require(json);
+    if (xpiPath) {
+      return resolve(getXPIManifest(xpiPath));
+    } else {
+      var cwd = process.cwd();
+      var json = path.join(cwd, "package.json");
+      var manifest = {};
+      try {
+        manifest = require(json);
+      }
+      catch (e) {}
+      return resolve(manifest);
     }
-    catch (e) {}
-    return resolve(manifest);
-  })
+  });
 }
 exports.getManifest = getManifest;
+
+
+function getXPIManifest(xpiPath) {
+  return when.promise(function(resolve, reject) {
+    tmp.dir(function(err, tmpPath, removeTmpDir) {
+      if (err) {
+        return reject(err);
+      }
+      fs.createReadStream(xpiPath)
+        .on('error', reject)
+        .pipe(unzip.Extract({path: tmpPath}))
+        .on('error', reject)
+        .on('close', function() {
+          resolve({
+            path: tmpPath,
+            remove: removeTmpDir,
+          });
+        });
+
+    }, {
+      // This allows us to remove a non-empty tmp dir.
+      unsafeCleanup: true,
+    });
+  })
+  .then(function(tmpDir) {
+    var manifest = {};
+    try {
+      manifest = require(path.join(tmpDir.path, "package.json"));
```

Comment:
Couldn't this move into the `close` handler and have the cleanup happen there? Doesn't really matter I guess but I was expecting the cleanup to be in there.